src/um/ipexport.rs

```diff
@@ -0,0 +1,174 @@
+// Licensed under the Apache License, Version 2.0
+// <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// All files in the project carrying such notice may not be copied, modified, or distributed
+// except according to those terms.
+// #include <winapifamily.h>
+// #include <in6addr.h>
+// #include <inaddr.h>
+use shared::basetsd::ULONG64;
+use shared::in6addr::in6_addr;
+use shared::ntdef::{INT, PUCHAR, PVOID, UCHAR, ULONG, USHORT, WCHAR};
+pub const MAX_ADAPTER_NAME: usize = 128;
+pub const MAX_OPT_SIZE: usize = 40;
+pub type IPAddr = ULONG;
```

Comment:
I wanted to point out that I think this definition of IPAddr may cause issues. This allows one to write 

```rust
let rust_ip = Ipv4Addr::new(1,2,3,4);
let win_ip: IPAddr = rust_ip.into();
```

However this actually swaps the byteorder. A packet sent to the `win_ip` value will go to 4.3.2.1 instead of 1.2.3.4.

Reply:
Regardless, it is the correct definition, and you just have to be really careful about byte order when working with windows API.

Reply:
I don't disagree that it is "correct", but perhaps it is worth adding a small doc comment or something?

I was (until just now) actually under the impression that IPAddr should have been an alias for IN_ADDR, considering everywhere I see IPAddr in MS documentation, it links to IN_ADDR instead. Looking at the header file, I see I was incorrect, though docs do say that the types are compatible.

I'm just concerned that this is a very easy mistake to make outside of an unsafe block and there is no documentation that states this is a problem.

Reply:
> ```rust
> let rust_ip = Ipv4Addr::new(1,2,3,4);
> let win_ip: IPAddr = rust_ip.into();
> ```
> 
> However this actually swaps the byteorder. A packet sent to the `win_ip` value will go to 4.3.2.1 instead of 1.2.3.4.

> I was (until just now) actually under the impression that IPAddr should have been an alias for IN_ADDR, considering everywhere I see IPAddr in MS documentation, it links to IN_ADDR instead. Looking at the header file, I see I was incorrect, though docs do say that the types are compatible.

The issue has nothing to do with MSDN docs. You may check the rust docs and find out things like this:

```rust
impl From<Ipv4Addr> for u32
fn from(ip: Ipv4Addr) -> u32
Converts an Ipv4Addr into a host byte order u32.
```

`rust_ip.into()` produces a host byte order u32. You're likely working on a little-endian machine.

Reply:
I'll accept that nothing will be done to address this; so be it. I still feel very uncomfortable about the fact that

```rust
let ip = IPAddr::from(Ipv4Addr::new(...));
```

looks like completely correct code, but it is not. It's not obvious that u32 is even involved. Nor does any bit of documentation make this obvious. The *only* way you know this is invalid is to actually look at the definition of IPAddr and see that it is actually u32.

> you're likely working on a little-endian machine.

As far as I know, Windows only runs on little endian so this is a potential issue for every user, which is why I even thought bring this up. I still think a small doc comment would go a long way, but again I accept that this won't happen.